metricq_manager/queue_manager.py

```diff
@@ -0,0 +1,526 @@
+    Exchange,
+    ExchangeType,
+    Message,
+    Queue,
+    RobustChannel,
+    RobustConnection,
+)
+from aio_pika.exceptions import AMQPChannelError, ChannelClosed
+from aiocouch.database import Database, NotFoundError
+from metricq import get_logger
+
+from .config_parser import ConfigParser
+
+logger = get_logger(__name__)
+
+Seconds = Union[int, float]
+
+HreqQueue = Queue
+DataQueue = Queue
+
+DataExchange = Exchange
+HistoryExchange = Exchange
+
+Metric = str
+DataQueueName = str
+HreqQueueName = str
+
+
+class SinkUnbindFailed(RuntimeError):
+    pass
+
+
+class QueueManager:
+    """The queue manager.
+
+    Args:
+        data_connection:
+            An open AMQP connection.
+            All queue operations are performed on the vhost to which it connects.
+        config_db:
+            An open connection to the CouchDB database that contains the per-client configuration.
+        channel:
+            An open channel on :code:`data_connection`.
+            If omitted, a new channel is opened on demand.
+    """
+
+    def __init__(
+        self,
+        *,
+        data_connection: RobustConnection,
+        config_db: Database,
+    ):
+        self.data_connection = data_connection
+        self.config = config_db
+
+        # The exchanges to which metric data are sent.
+        # The need to be declared with :meth:`declare_exchanges` before use.
+        self._data_exchange: Optional[DataExchange] = None
+        self._history_exchange: Optional[HistoryExchange] = None
+
+    async def close(self):
+        """Shutdown this queue manager.
+
+        This closes the data connection.
+        """
+        logger.debug("Closing data connection {!r}", self.data_connection)
+        await self.data_connection.close()
+
+    @asynccontextmanager
+    async def temporary_channel(self, reuse: Optional[RobustChannel] = None):
+        if reuse is not None:
+            logger.debug("Reusing channel {!r} as temporary channel", reuse)
+            yield reuse
+            return
+
+        channel: RobustChannel = await self.data_connection.channel()
+        logger.debug("Opened temporary channel {!r}", channel)
+
+        try:
+            yield channel
+        except ChannelClosed:
+            logger.warning("Temporary channel {!r} closed unexpectedly!", channel)
+        finally:
+            logger.debug("Closing temporary channel {!r}", channel)
+            try:
+                await channel.close()
+            except AMQPChannelError as e:
+                logger.warning("Failed to close temporary channel {!r}: {}", channel, e)
+
+    async def declare_queue(
+        self,
+        name: str,
+        arguments: Dict[str, Any],
+        auto_delete: bool = False,
+        durable: bool = False,
+        channel: Optional[RobustChannel] = None,
+    ) -> Queue:
+        """Declare a new queue."""
+
+        async with self.temporary_channel(reuse=channel) as channel:
+            return await channel.declare_queue(
```

Comment:
Well... no. The ominous `Writer is None` error occurs then. I could probably make a type alias so that methods like this are only ever called by channels created with `temporary_channel()`.

Reply:
Oh, now I see. I misunderstood your comment. You are talking about what happens to the queue after the temporary channel is closed at the end of the `with` block. Yes, this method (and `_declare_exchange`) should probably not open their own channels and instead take `channel` as a required argument.

Reply:
Check out 171405e.